lib/cid.ex

```diff
@@ -1,38 +1,87 @@
 defmodule Cid do
   @moduledoc """
-  Returns a SHA512 transformed to Base64, remove ambiguous chars then sub-string
+  Provides a way for a user to turn a String, Map or Struct into a CID that
+  is identical to one what will be returned from IPFS if the same data is
+  added.
+
+  Currently only produces a default v1 CID.
+  Currently only uses the "raw" codec
+  Data provided must be under 256Kb in order for the CID to match the one
+  returned by IPFS
   """
 
   @doc """
-  make/2 create a SHA512 hash from the given input and return the require length
-  note: we remove "ambiguous" characters so _humans_ can type the hash without
-  getting "confused" this might not be required, but is to match the original
-  "Hits" implementation.
+  Returns a CID that identical to one returned by IPFS if given the same data.
+  Can take a String, Map or Struct as an argument.
 
-  ## Parameters
+  ## Examples
 
-  - input: String the string to be hashed.
-  - length: Number the length of string required
+    iex> Cid.cid("hello")
+    "zb2rhcc1wJn2GHDLT2YkmPq5b69cXc2xfRZZmyufbjFUfBkxr"
 
-  Returns String hash of desired length.
+    iex> Cid.cid(%{key: "value"})
+    "zb2rhkN6szWhAmBFjjP8RSczv2YVNLnG1tz1Q7FyfEp8LssNZ"
   """
-  def make(input) when is_map(input) do
-    input |> stringify_map_values |> make
+  def cid(value) do
+    value
+    |> create_multihash()
+    |> create_cid()
+  end
+
+  # if create_multihash is called with a struct, the struct is converted into a
+  # map and then create_multihash is called again
+  defp create_multihash(%_{} = struct) do
+    struct
+    |> Map.from_struct()
+    |> create_multihash()
   end
 
-  def make(input, length \\ 32) do
-    hash1 = :crypto.hash(:sha512, input)
-    {:ok, <<_multihash_code, _length, hash2::binary>>} = Multihash.encode(:sha2_512, hash1)
+  # if create_multihash is called with a map the map is converted into a JSON
+  # string and then create_multihash is called again
+  defp create_multihash(map) when is_map(map) do
+    map
+    |> Jason.encode!()
+    |> create_multihash()
+  end
+
+  # if create_multihash is called with a string, the string has a new line added
+  # to the end (as that's what IPFS appears to be doing based on tests), then
+  # the string is converted into a multihash
+  defp create_multihash(str) when is_binary(str) do
+    str = add_new_line(str)
+    digest = :crypto.hash(:sha256, str)
```

Comment:
Explaining in the comments that this is an erlang function would be good for beginners